lucene/core/src/java/org/apache/lucene/geo/Circle2D.java

```diff
@@ -0,0 +1,463 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.geo;
+
+import org.apache.lucene.index.PointValues.Relation;
+import org.apache.lucene.util.SloppyMath;
+
+/**
+ * 2D circle implementation containing spatial logic.
+ */
+class Circle2D implements Component2D {
+
+  private final DistanceCalculator calculator;
+
+  private Circle2D(DistanceCalculator calculator) {
+    this.calculator = calculator;
+  }
+
+  @Override
+  public double getMinX() {
+    return calculator.getMinX();
+  }
+
+  @Override
+  public double getMaxX() {
+    return calculator.getMaxX();
+  }
+
+  @Override
+  public double getMinY() {
+    return calculator.getMinY();
+  }
+
+  @Override
+  public double getMaxY() {
+    return calculator.getMaxY();
+  }
+
+  @Override
+  public boolean contains(double x, double y) {
+    return calculator.contains(x, y);
+  }
+
+  @Override
+  public Relation relate(double minX, double maxX, double minY, double maxY) {
+    if (calculator.disjoint(minX, maxX, minY, maxY)) {
+      return Relation.CELL_OUTSIDE_QUERY;
+    }
+    if (calculator.within(minX, maxX, minY, maxY)) {
+      return Relation.CELL_CROSSES_QUERY;
+    }
+    return calculator.relate(minX, maxX, minY, maxY);
+  }
+
+  @Override
+  public Relation relateTriangle(double minX, double maxX, double minY, double maxY,
+                                 double ax, double ay, double bx, double by, double cx, double cy) {
+    if (calculator.disjoint(minX, maxX, minY, maxY)) {
+      return Relation.CELL_OUTSIDE_QUERY;
+    }
+    if (ax == bx && bx == cx && ay == by && by == cy) {
+      // indexed "triangle" is a point: shortcut by checking contains
+      return contains(ax, ay) ? Relation.CELL_INSIDE_QUERY : Relation.CELL_OUTSIDE_QUERY;
+    } else if (ax == cx && ay == cy) {
+      // indexed "triangle" is a line segment: shortcut by calling appropriate method
+      return relateIndexedLineSegment(ax, ay, bx, by);
+    } else if (ax == bx && ay == by) {
+      // indexed "triangle" is a line segment: shortcut by calling appropriate method
+      return relateIndexedLineSegment(bx, by, cx, cy);
+    } else if (bx == cx && by == cy) {
+      // indexed "triangle" is a line segment: shortcut by calling appropriate method
+      return relateIndexedLineSegment(cx, cy, ax, ay);
+    }
+    // indexed "triangle" is a triangle:
+    return relateIndexedTriangle(minX, maxX, minY, maxY, ax, ay, bx, by, cx, cy);
+  }
+
+  @Override
+  public WithinRelation withinTriangle(double minX, double maxX, double minY, double maxY,
+                                       double ax, double ay, boolean ab, double bx, double by, boolean bc, double cx, double cy, boolean ca) {
+    // short cut, lines and points cannot contain this type of shape
+    if ((ax == bx && ay == by) || (ax == cx && ay == cy) || (bx == cx && by == cy)) {
+      return WithinRelation.DISJOINT;
```

Comment:
couldn't it be NOTWITHIN in some cases?

Reply:
Strictly speaking yes but a point or a line would never return CANDIDATE for a circle. I guess I am using disjoint because is the cheaper answer (the shape is actually ignored) but I agree this is probably wrong.